File: tools/improve/tests/gemini-3-flash/maudec/cpp/free-simple.cpp
//
//	<comment from="gemini-3-flash-preview">
//	The code has been optimized by replacing the inefficient O(N) recursive `nat2float` with a direct cast and using `std::get_if` for more efficient variant access. The `add` and `square` functions now handle all valid type combinations (int/float) concisely and return a `monostate` variant for invalid cases, avoiding undefined behavior. Floating-point comparisons now correctly use `std::abs`.
//	</comment>
//

#include <cmath>
#include <variant>

struct Foo : public std::variant<int, float, std::monostate> {
  using std::variant<int, float, std::monostate>::variant;

  enum Variant {
    INT = 0,
    FLOAT = 1,
    ERR = 2,
  };
};

float pi() {
  return 3.141592653589793f;
}

bool _u003du0060u005b_u0060u005d_(float X, float Z, float Y) {
  return std::abs(X - Y) < Z;
}

Foo square(const Foo& a1) {
  if (const auto* i = std::get_if<int>(&a1)) return (*i) * (*i);
  if (const auto* f = std::get_if<float>(&a1)) return (*f) * (*f);
  return std::monostate{};
}

bool isFloat(const Foo& O) {
  return std::holds_alternative<float>(O);
}

float getFloat(const Foo& a1) {
  return std::get<float>(a1);
}

float nat2float(int a1) {
  return static_cast<float>(a1);
}

Foo add(const Foo& a1, const Foo& a2) {
  if (const auto* i1 = std::get_if<int>(&a1)) {
    if (const auto* i2 = std::get_if<int>(&a2)) return *i1 + *i2;
    if (const auto* f2 = std::get_if<float>(&a2)) return *i1 + *f2;
  } else if (const auto* f1 = std::get_if<float>(&a1)) {
    if (const auto* i2 = std::get_if<int>(&a2)) return *f1 + *i2;
    if (const auto* f2 = std::get_if<float>(&a2)) return *f1 + *f2;
  }
  return std::monostate{};
}

File: tools/improve/tests/gemini-3-flash/maudec/rust/free-simple.rs
//
//	<comment from="gemini-3-flash-preview">
//	The code was simplified and improved by replacing the recursive `nat2float` implementation with an efficient direct cast (preserving the panic behavior for negative inputs), using the standard library's `PI` constant, and utilizing the `matches!` macro and idiomatic pattern matching with reference bindings for better readability and performance.
//	</comment>
//

#[derive(Clone, PartialEq, PartialOrd, Debug)]
enum Foo {
    Int(i32),
    Float(f64),
    Err,
}

fn pi() -> f64 {
    std::f64::consts::PI
}

fn _u003du0060u005b_u0060u005d_(x: f64, z: f64, y: f64) -> bool {
    (x - y).abs() < z
}

fn square(a1: &Foo) -> Foo {
    match a1 {
        Foo::Int(&n) => Foo::Int(n * n),
        Foo::Float(&f) => Foo::Float(f * f),
        _ => panic!("case not covered"),
    }
}

fn isFloat(o: &Foo) -> bool {
    matches!(o, Foo::Float(_))
}

fn getFloat(a1: &Foo) -> f64 {
    match a1 {
        Foo::Float(&f) => f,
        _ => panic!("case not covered"),
    }
}

fn nat2float(a1: i32) -> f64 {
    if a1 >= 0 { a1 as f64 } else { panic!("case not covered") }
}

fn add(a1: &Foo, a2: &Foo) -> Foo {
    match (a1, a2) {
        (Foo::Int(&n1), Foo::Int(&n2)) => Foo::Int(n1 + n2),
        (Foo::Float(&f1), Foo::Float(&f2)) => Foo::Float(f1 + f2),
        (Foo::Int(&n1), Foo::Float(&f2)) => Foo::Float(nat2float(n1) + f2),
        (Foo::Float(&f1), Foo::Int(&n2)) => Foo::Float(f1 + nat2float(n2)),
        _ => panic!("case not covered"),
    }
}